walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,55 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> existingEvents = new ArrayList<Event>(events);
+    List<TimeRange> possibleSchedule = new ArrayList<TimeRange>();
+
+    Collections.sort(existingEvents);
+    int curStartTime = 0;
+    TimeRange eventTimeRange;
+    boolean contains;
+
+    for (Event event : existingEvents) {
+      contains = false;
+      for (String participant : request.getAttendees()) {
+        if (event.getAttendees().contains(participant)) {
+          contains = true;
+          break;
+        }
+      }
+      if (!contains) {
+        continue;
+      }
+      
+      eventTimeRange = event.getWhen();
+      if (eventTimeRange.start() <= curStartTime) {
+        if (eventTimeRange.end() <= curStartTime) {
+          continue;
+        }
+        curStartTime = eventTimeRange.end();
+        continue;
+      }
+
+      if (eventTimeRange.start() - curStartTime >= request.getDuration()) {
+        possibleSchedule.add(TimeRange.fromStartDuration(curStartTime, eventTimeRange.start() - curStartTime));
+      }
+      curStartTime = eventTimeRange.end();
+
+    }
+
+    if (curStartTime != TimeRange.END_OF_DAY) {
```

Comment:
Can the if-statement-block potentially be refactored into a smaller function?

Reply:
Are you referring to the various if blocks that are used to handle each cases? I did not refactor them out to separate functions because they are simple enough. Or are you asking me to refactor it to increase readability?

Reply:
There are multiple ways things can be refactored:
* To extract the if-condition expression to separate method
* To extract the whole if-block to separate method

It definitely looks simple enough, especially when you are the writer of the code.
But cognitively for other code readers, it is more difficult to comprehend.
It can help to explain your code better, encapsulate logic without revealing the details, easier maintenance, etc.

For this, one suggestion you can give a try for illustration: `addToScheduleIfMeetingFitsAtEndOfDay(...)`

Reply:
Refactored the if-else statements.